blog/2024-12-03-configuration_preview.md

```diff
@@ -0,0 +1,137 @@
+- Configurations are more easily modularized.
+- If a key name or default value changes in the future, it will only be a breaking change if it was a value that had been overridden. All other values will update seamlessly when you install new Nushell releases.
+
+  ::: note
+  This may be an advantage or disadvantage in some situations. For instance, at some point, we plan to switch the default history format to SQLite. When that change occurs in Nushell, it will automatically be changed for all users who hadn't overridden the value. That's a positive change for most users, as they'll automatically be switched to the more advanced format when they upgrade to that (as yet unknown) release, but that change may not be desirable for some users.
+
+  Of course, these users can always simply override that value when and if the default changes.
+  :::
+
+Not _every_ default value is introspectable. The following Nushell internals are no longer set (by default) in `config.nu` and will not be automatically populated:
+
+- `keybindings`
+- `menus`
+
+However, the functionality behind them has always been handled internally by Nushell and Reedline. Only user-defined keybindings and menus should (as best-practice) be specified in the `$env.config`.
+
+### Finding Overridden Values
+
+To identify which values your current configuration has changed from the defaults, run the following in the current build (or 0.101 when available):
+
+```nu
+let defaults = nu -n -c "$env.config = {}; $env.config | reject color_config keybindings menus | to nuon" | from nuon | transpose key default
+let current = $env.config | reject color_config keybindings menus | transpose key current
+$current | merge $defaults | where $it.current != $it.default
+```
+
+These are the values that you should migrate to your updated `config.nu`.
+
+::: note
+In the above example, `nu` (without a path) needs to point to a 0.101 or higher release in order for this to work. This should be the normal result, but users who are temporarily running from a compiled build (e.g. `./target/release/nu`) may need to adjust the command.
+:::
+
+Also examine:
+
+- Any theme/styling in `$env.config.color_config` and add those settings if desired.
+- Your personalized keybindings in `$env.config.keybindings`. Note that most (perhaps all) of the keybindings in the older default configuration files were simply examples that replicated built-in capabilities and did not change any Nushell functionality.
+- Any personalized menus in `$env.config.menus`. As with keybindings, you do not need to copy over examples.
+
+### Setting Values in the New Config
+
+Rather than defining a monolithic **`$env.config = { ... all values }`** as in the past, just create one entry for each setting you wish to **_override_**. For example:
+
+```nu
+$env.config.show_banner = false
+$env.config.buffer_editor = "code"
+
+$env.history.file_format = "sqlite"
+$env.history.max_size: 1_000_000
+$env.history.isolation = true
```

Comment:
aren't all 3 of these supposed to be `$env.config.history`...?

Reply:
I had them there and then reversed it.  The point is that you can use the complete record, but you can also override individual settings.  In this case `$env.history.sync_on_enter` isn't needed since the user wants to keep the default value.

Let me see if I can make it more clear ...

Reply:
Oh, I misunderstood - I see the issue now.  Fixed in latest release.